app/api.js

```diff
@@ -91,47 +91,102 @@ export async function setPassword(id, owner_token, keychain) {
   return response.ok;
 }
 
-export function uploadFile(
+function asyncInitWebSocket(server) {
+  return new Promise(resolve => {
+    const ws = new WebSocket(server);
+    ws.onopen = () => {
+      resolve(ws);
+    };
+  });
+}
+
+async function upload(
+  ws,
+  stream,
+  streamInfo,
+  metadata,
+  verifierB64,
+  keychain,
+  onprogress,
+  canceller
+) {
+  const metadataHeader = arrayToB64(new Uint8Array(metadata));
+  const fileMeta = {
+    fileMetadata: metadataHeader,
+    authorization: `send-v1 ${verifierB64}`
+  };
+
+  function listenForResponse() {
+    return new Promise((resolve, reject) => {
+      ws.addEventListener('message', function(msg) {
+        const response = JSON.parse(msg.data);
+        if (response.error) {
+          reject(response.error);
+        } else {
+          resolve({
+            url: response.url,
+            id: response.id,
+            ownerToken: response.owner
+          });
+        }
+      });
+    });
+  }
+
+  const resPromise = listenForResponse();
+  ws.send(JSON.stringify(fileMeta));
```

Comment:
What happens if the server replies with an error? It looks like we still try to send the whole file. One way to stop early is to use a similar technique to the cancelled flag. Set an `errored` flag in `listenForResponse` and check it in the loop.